### PR TITLE
Reindexes a single resource

### PR DESCRIPTION
**Pull request by rotodd**
## Description

Prior to this, we were relying on the `UpsertResource` stored procedure to update a resource's search indices in SQL. This created a new version of every resource that was reindexed.

This PR creates a new stored proc for reindexing that:

1. Updates the search parameter hash value in the dbo.Resource table
2. Deletes all of the resource's entries in the search parameter index tables
3. Inserts new entries into the search parameter index tables

Note: we could improve the performance of this by only updating the search indices that are out of date. [AB#79648](https://microsofthealth.visualstudio.com/Health/_workitems/edit/79648) is being used to track this.

## Related issues

Addresses: [AB#79121](https://microsofthealth.visualstudio.com/Health/_workitems/edit/79121)
Coming up next: [AB#73395](https://microsofthealth.visualstudio.com/Health/_workitems/edit/73395)

## Testing

Adds a check to ensure that the version of resource does not change in existing reindex tests.

## FHIR Team Checklist
- ✅ **Update the title** of the PR to be succinct and less than 50 characters
- ✅ **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- ✅ Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- ✅ Tag the PR with **Azure API for FHIR** if this will release to the managed service
- ✅ Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Features/Persistence/IFhirDataStore.cs
Comment (rotodd) [inline src/Microsoft.Health.Fhir.Core/Features/Persistence/IFhirDataStore.cs:26]: @feordin I removed this, since it has no usages and it is easier to update the search parameter hash at the same time as the indices - is there anything I'm missing/any reason to keep it?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:776]: @johnstairs this lock is the same one that we are using in the `UpsertResource` stored proc. 

To confirm my understanding, this is only locking certain access to the `dbo.Resource` table, right? If so, is it alright that we are not blocking read/write access to all the index tables during an upsert or reindex operation?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:783]: @feordin I am passing the etag into the `ReindexResource` stored proc to perform an additional version check here - but it doesn't seem all that necessary. Is there a reason why this is needed on the Cosmos side of things?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:795]: Did you consider the `MERGE` statement instead of `DELETE` followed by `INSERT`? In most reindexing cases, most of the existing indexes would end up being the same. A merge _might_ be more efficient here.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs
Comment (rotodd) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:320]: Remove this if the version check isn't needed [here](https://github.com/microsoft/fhir-server/pull/1682/files#r579401925).

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (johnstairs) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:776]: All reads and writes that you do within a transaction will place different kinds of locks.

Since we are using RCSI, reads will not be blocked by this operation. For writes, concurrent calls to this sproc and  UpsertResource would block each other if accessing the same resource, but that should be by design. Updating separate resources should be fine.

File: src/Microsoft.Health.Fhir.Core/Features/Persistence/IFhirDataStore.cs
Comment (feordin) [reply src/Microsoft.Health.Fhir.Core/Features/Persistence/IFhirDataStore.cs:26]: We will only end up using this is if we implement the performance enhancement which compares preivous indices to the newly extracted indices, and then determine that the only update required is the hash value.  We can add this method back in if we ever implement that.  So it is fine to remove it.

Comment (rotodd) [reply src/Microsoft.Health.Fhir.Core/Features/Persistence/IFhirDataStore.cs:26]: Ah, I see - thank you! I'll make a note about it in [AB#79648](https://microsofthealth.visualstudio.com/Health/_workitems/edit/79648).

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:795]: Good suggestion - looking into this now. Thanks!

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (feordin) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:783]: I believe it is here in case the resource is changed between the time it was initially queried by the reindexjob, and the time the process finishes extracting the new search indices and then calls this stored procedure to update the values.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:795]: As discussed, I will consider this when writing the stored proc for batch updates in [AB#73395](https://microsofthealth.visualstudio.com/Health/_workitems/edit/73395), since that is where we will benefit most from performance gains.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql:782]: As discussed, it would be good if the reindex job worker could pick up these resources and attempt to reindex them again instead of returning an error. Created [AB#79925](https://microsofthealth.visualstudio.com/Health/_workitems/edit/79925) and added some missing exception handling in 9122322 in the meantime.

**Review by feordin — APPROVED** [review]

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/8.diff.sql]: Thanks for the reply!

I felt like I still needed a refresher on this, so I read through [this doc](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server). Jotting down my takeaways here for my future self:

- Read Committed Snapshot Isolation (RCSI) ensures that data read within a transaction will never reflect changes made by other simultaneous transactions
- With RSCI, the data row versions are recorded at the beginning of a transaction, and if a transaction attempts to commit an update to a row that was changed after the transaction began, the transaction is rolled back, and an error is raised
- Lock hints like `UPDLOCK` can reduce the chance of update conflicts in long running transactions by blocking other transactions from attempting to update a specific set of rows before the transaction completes

The stored proc above begins a new transaction which has RSCI enabled. The lock is used to block other transactions from updating the row in `dbo.Resource` for the resource being reindexed. In theory, the data in the index tables relevant to this resource could be modified over the course of the transaction. If that were the case, the transaction would fail to commit, and an error would be returned.

@johnstairs please let me know if I have misunderstood something here
@feordin it sounds like this would be another similar scenario where we would like to reattempt to reindex a resource

**Pull request merged**
